## Film Strategy Study
This notebook is being used to explore the data available through the im.db database.

In [1]:
import pandas as pd
import sqlite3

In [2]:
!pwd

/c/Users/NM/Documents/Flatiron/phase_1/Film-Strategy/Study


In [3]:
conn = sqlite3.connect('../zippedData/im.db')

**Let's look at the db schema.**

In [4]:
query = """
SELECT *
FROM sqlite_master
"""
pd.read_sql_query(query, conn)

type           name       tbl_name  rootpage  \
0  table   movie_basics   movie_basics         2   
1  table      directors      directors         3   
2  table      known_for      known_for         4   
3  table     movie_akas     movie_akas         5   
4  table  movie_ratings  movie_ratings         6   
5  table        persons        persons         7   
6  table     principals     principals         8   
7  table        writers        writers         9   

                                                 sql  
0  CREATE TABLE "movie_basics" (\n"movie_id" TEXT...  
1  CREATE TABLE "directors" (\n"movie_id" TEXT,\n...  
2  CREATE TABLE "known_for" (\n"person_id" TEXT,\...  
3  CREATE TABLE "movie_akas" (\n"movie_id" TEXT,\...  
4  CREATE TABLE "movie_ratings" (\n"movie_id" TEX...  
5  CREATE TABLE "persons" (\n"person_id" TEXT,\n ...  
6  CREATE TABLE "principals" (\n"movie_id" TEXT,\...  
7  CREATE TABLE "writers" (\n"movie_id" TEXT,\n  ...

**Examining the movie_basics table**

In [5]:
query = """
SELECT *
FROM movie_basics
GROUP BY movie_id
-- LIMIT 5
"""
pd.read_sql_query(query,conn)

movie_id                                primary_title  \
0       tt0063540                                    Sunghursh   
1       tt0066787              One Day Before the Rainy Season   
2       tt0069049                   The Other Side of the Wind   
3       tt0069204                              Sabse Bada Sukh   
4       tt0100275                     The Wandering Soap Opera   
...           ...                                          ...   
146139  tt9916538                          Kuambil Lagi Hatiku   
146140  tt9916622  Rodolpho Teóphilo - O Legado de um Pioneiro   
146141  tt9916706                              Dankyavar Danka   
146142  tt9916730                                       6 Gunn   
146143  tt9916754               Chico Albuquerque - Revelações   

                                     original_title  start_year  \
0                                         Sunghursh        2013   
1                                   Ashad Ka Ek Din        2019   
2                        The Other Side of the Wind        2018   
3                                   Sabse Bada Sukh        2018   
4                             La Telenovela Errante        2017   
...                                             ...         ...   
146139                          Kuambil Lagi Hatiku        2019   
146140  Rodolpho Teóphilo - O Legado de um Pioneiro        2015   
146141                              Dankyavar Danka        2013   
146142                                       6 Gunn        2017   
146143               Chico Albuquerque - Revelações        2013   

        runtime_minutes                genres  
0                 175.0    Action,Crime,Drama  
1                 114.0       Biography,Drama  
2                 122.0                 Drama  
3                   NaN          Comedy,Drama  
4                  80.0  Comedy,Drama,Fantasy  
...                 ...                   ...  
146139            123.0                 Drama  
146140              NaN           Documentary  
146141              NaN                Comedy  
146142            116.0                  None  
146143              NaN           Documentary  

[146144 rows x 6 columns]

**movie_id is probably the value needed to connect to other tables such as director, writer, etc.**
**Let's look at the director table**

In [6]:
query = """
SELECT *
FROM directors
LIMIT 5
"""
pd.read_sql_query(query, conn)

movie_id  person_id
0  tt0285252  nm0899854
1  tt0462036  nm1940585
2  tt0835418  nm0151540
3  tt0835418  nm0151540
4  tt0878654  nm0089502

**Presumably, for the movie represented by movie_id, the person_id points to detail about the director. Implication is that person_id is probably to be found in the table called persons.**

In [7]:
query = """
SELECT *
FROM persons
LIMIT 5
"""
pd.read_sql_query(query, conn)

person_id       primary_name birth_year death_year  \
0  nm0061671  Mary Ellen Bauder       None       None   
1  nm0061865       Joseph Bauer       None       None   
2  nm0062070         Bruce Baum       None       None   
3  nm0062195       Axel Baumann       None       None   
4  nm0062798        Pete Baxter       None       None   

                                 primary_profession  
0         miscellaneous,production_manager,producer  
1        composer,music_department,sound_department  
2                        miscellaneous,actor,writer  
3  camera_department,cinematographer,art_department  
4  production_designer,art_department,set_decorator

**What does movie_ratings table contain?**

In [8]:
query = """

SELECT *
FROM movie_ratings
LIMIT 5
"""
pd.read_sql_query(query, conn)

movie_id  averagerating  numvotes
0  tt10356526            8.3        31
1  tt10384606            8.9       559
2   tt1042974            6.4        20
3   tt1043726            4.2     50352
4   tt1060240            6.5        21

**So, movie_ratings table provides the average rating across the given number of votes for the movie represented by movie_id.**

**Let's look at movies_akas to see if we can correlate with language**

In [20]:
query = """
SELECT *
FROM movie_akas
WHERE region = 'US'
LIMIT 20

"""
pd.read_sql_query(query, conn)

movie_id  ordering                                       title region  \
0    tt0369610        21                           Jurassic World 3D     US   
1    tt0369610        29                              Jurassic World     US   
2    tt0369610         2                                    Ebb Tide     US   
3    tt0369610        36                            Jurassic Park IV     US   
4    tt0369610        44                             Jurassic Park 4     US   
5    tt0401729        11                         John Carter of Mars     US   
6    tt0401729        12                          A Princess of Mars     US   
7    tt0401729         2                                 John Carter     US   
8    tt0401729         5                                     Barsoom     US   
9    tt1014759        23  Alice in Wonderland: An IMAX 3D Experience     US   
10   tt1014759        37                         Alice in Wonderland     US   
11   tt1014759         6                                       Alice     US   
12  tt10379338         1                              Guys Night Out     US   
13  tt10379338         3                            Random Night Out     US   
14  tt10446418         3                                    The Icon     US   
15   tt1180174         1                                     Liberty     US   
16   tt1182315         4                           Dark Buenos Aires     US   
17   tt1189904         1                             Small Town Hero     US   
18   tt1194173         2                                     Marcher     US   
19   tt1194173         9                           The Bourne Legacy     US   

   language        types                  attributes  is_original_title  
0      None         None                 3-D version                0.0  
1      None         None                        None                0.0  
2      None         None          fake working title                0.0  
3      None      working                        None                0.0  
4      None         None  informal alternative title                0.0  
5      None      working                        None                0.0  
6      None      working                        None                0.0  
7      None         None                        None                0.0  
8      None         None          fake working title                0.0  
9      None         None                IMAX version                0.0  
10     None         None                        None                0.0  
11     None         None       original script title                0.0  
12     None  imdbDisplay                        None                0.0  
13     None         None                        None                0.0  
14     None  alternative                        None                0.0  
15     None         None                        None                0.0  
16     None  imdbDisplay                        None                0.0  
17     None         None                        None                0.0  
18     None         None          fake working title                0.0  
19     None         None                        None                0.0

In [21]:
query = """
SELECT *
FROM movie_akas
WHERE region = 'IN'
"""
m_akas_df = pd.read_sql_query(query, conn)
m_akas_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8435 entries, 0 to 8434
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   movie_id           8435 non-null   object 
 1   ordering           8435 non-null   int64  
 2   title              8435 non-null   object 
 3   region             8435 non-null   object 
 4   language           1376 non-null   object 
 5   types              563 non-null    object 
 6   attributes         685 non-null    object 
 7   is_original_title  8434 non-null   float64
dtypes: float64(1), int64(1), object(6)
memory usage: 527.3+ KB


**Too many nulls in Language and randomly browsing with regions, there is no clean way to distinguish English movies from other languages. However, it might be worth investigating the two very large producers of movies - US and India. It may be pertinent to Microsoft which has a large presence in both of these markets.**

In [ ]:
Used to randomly look at movie records by manually providing a movie_id.

In [ ]:
query = """
SELECT *
FROM movie_basics
WHERE movie_id = 'tt10356526'
LIMIT 5
"""
pd.read_sql_query(query,conn)

**Let's join the movie name and info with the director and ratings.**

In [ ]:
query = """
SELECT DISTINCT *
FROM movie_basics
LEFT JOIN directors
ON movie_basics.movie_id = directors.movie_id
LEFT JOIN persons
ON directors.person_id = persons.person_id
LEFT JOIN movie_ratings
ON movie_basics.movie_id = movie_ratings.movie_id
LIMIT 10
"""
pd.read_sql_query(query, conn)

**Trying to be a little more judicious with the number of columns being selected.**
**Many of the top scoring movies seem to have relatively few votes.**
**Lets look at movies with at least 500 votes and limit display to first 100 rows.**

In [ ]:
query = """
SELECT DISTINCT mb.primary_title AS "Movie Name", mb.start_year AS "Release Year", p.primary_name AS "Director", mr.averagerating AS "Avg Rating", mr.numvotes AS "Number of Votes"

FROM movie_basics mb
    
    LEFT JOIN directors d
        ON mb.movie_id = d.movie_id
    LEFT JOIN persons p
        ON d.person_id = p.person_id
    LEFT JOIN movie_ratings mr
        ON mb.movie_id = mr.movie_id

WHERE mr.numvotes > 500

ORDER BY mr.averagerating DESC

LIMIT 100
"""
pd.read_sql_query(query, conn)

**Capturing the query in a data frame without limit on number of rows**

In [ ]:
query = """
SELECT DISTINCT mb.primary_title AS "Movie Name", mb.start_year AS "Release Year", p.primary_name AS "Director", mr.averagerating AS "Avg Rating", mr.numvotes AS "Number of Votes"

FROM movie_basics mb
    
    LEFT JOIN directors d
        ON mb.movie_id = d.movie_id
    LEFT JOIN persons p
        ON d.person_id = p.person_id
    LEFT JOIN movie_ratings mr
        ON mb.movie_id = mr.movie_id

WHERE mr.numvotes > 500

ORDER BY mr.averagerating DESC

"""
imdb_df = pd.read_sql_query(query, conn)
len(imdb_df)

**Out of the original 146,144 rows in the movie_basics table, only 15,595 rows meet this criteria.**

In [ ]:
imdb_df.head(100)

In [ ]:
imdb_df.info()

In [ ]:
imdb_df.loc[imdb_df['Director'] == 'Quentin Tarantino']['Movie Name']

In [ ]:
imdb_df['Director'].isnull().sum()

In [22]:
query = """
SELECT DISTINCT mb.primary_title AS "Movie Name", mb.start_year AS "Release Year", p.primary_name AS "Director", mr.averagerating AS "Avg Rating", mr.numvotes AS "Number of Votes"

FROM movie_basics mb
    
    LEFT JOIN directors d
        ON mb.movie_id = d.movie_id
    LEFT JOIN persons p
        ON d.person_id = p.person_id
    LEFT JOIN movie_ratings mr
        ON mb.movie_id = mr.movie_id

WHERE mr.numvotes > 500 AND p.primary_name IS Null

ORDER BY mr.averagerating DESC

"""
pd.read_sql_query(query, conn)

Movie Name  Release Year Director  Avg Rating  \
0                 On vam ne Dimon          2017     None         9.2   
1                  Burning Love 2          2014     None         8.0   
2                    Free to Play          2014     None         8.0   
3  Digimon Adventure Tri. 4: Loss          2017     None         7.7   
4   Horoob Etirari: Forced Escape          2017     None         6.0   
5                         Ecstasy          2011     None         4.3   
6           Rise of the Black Bat          2012     None         1.2   

   Number of Votes  
0             2721  
1              531  
2            12704  
3              521  
4              802  
5              721  
6              616